
Exploring Hackers News Posts
In this project, we'll compare two different types of posts from Hacker News, a popular site where technology related stories (or 'posts') are voted and commented upon. The two types of posts we'll explore begin with either Ask HN or Show HN.

Users submit Ask HN posts to ask the Hacker News community a specific question, such as "What is the best online course you've ever taken?" Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting.

We'll specifically compare these two types of posts to determine the following:

Do Ask HN or Show HN receive more comments on average?
Do posts created at a certain time receive more comments on average?
It should be noted that the data set we're working with was reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions.


Introduction
First, we'll read in the data and remove the headers

In [1]:
# Read in the data.
import csv

f = open('hacker_news.csv')
hn = list(csv.reader(f))
hn[:5]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

Extract the first row of data, and assign it to the variable headers.
Remove the first row from hn.
Display headers.
Display the first five rows of hn to verify that you removed the header row properly.

In [2]:
# Remove the headers.
headers = hn[0]
hn = hn[1:]
print(headers)
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


Create three empty lists called ask_posts, show_posts, and other_posts.
Loop through each row in hn.
Assign the title in each row to a variable named title.
Because the title column is the second column, you'll need to get the element at index 1 in each row.
Implement the following steps:
If the lowercase version of title starts with ask hn, append the row to ask_posts.
Else if the lowercase version of title starts with show hn, append the row to show_posts.
Else append to other_posts.
Check the number of posts in ask_posts, show_posts, and other_posts.

In [3]:
# Identify posts that begin with either `Ask HN` or `Show HN` and separate the data into different lists.
ask_posts = []
show_posts =[]
other_posts = []

for post in hn:
    title = post[1]
    if title.lower().startswith("ask hn"):
        ask_posts.append(post)
    elif title.lower().startswith("show hn"):
        show_posts.append(post)
    else:
        other_posts.append(post)
        
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))




1744
1162
17194


Find the total number of comments in ask posts and assign it to total_ask_comments.
Set total_ask_comments to 0.
Use a for loop to iterate over the ask posts.
Because the num_comments column is the fifth column in ask_posts, you'll need to get the element at index 4 in each row.
You'll also need to convert the value to an integer so that we can calculate the sum of all the comments.
Add this value to total_ask_comments.
Compute the average number of comments on ask posts and assign it to avg_ask_comments.
Print avg_ask_comments.
Find the total number of comments in show posts and assign it to total_show_comments.
Set total_show_comments to 0.
Use a for loop to iterate over the show posts.
Because the num_comments column is the fifth column in show_posts, you'll need to get the element at index 4 in each row.
You'll also need to convert the value to an integer so that we can calculate the sum of all the comments.
Add this value to total_show_comments.
Compute the average number of comments on show posts and assign it to avg_show_comments.
Print avg_show_comments.
Do show posts or ask posts receive more comments on average? Write a markdown cell explaining your findings.

In [4]:
# Calculate the average number of comments `Ask HN` posts receive.
total_ask_comments = 0

for post in ask_posts:
    total_ask_comments += int(post[4])
    
avg_ask_comments = total_ask_comments / len(ask_posts)
print(avg_ask_comments)

    

14.038417431192661


In [5]:
total_show_comments = 0

for post in show_posts:
    total_show_comments += int(post[4])
    
avg_show_comments = total_show_comments / len(show_posts)
print(avg_show_comments)

10.31669535283993


On average, ask posts in our sample receive approximately 14 comments, whereas show posts receive approximately 10. Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts.

Recall that we can use the datetime.strptime() constructor to parse dates stored as strings and return datetime objects:

date_1_str = "December 24, 1984"
date_1_dt = dt.datetime.strptime(date_1_str, "%B %d, %Y")

Import the datetime module as dt.
Create an empty list and assign it to result_list. This will be a list of lists.
Iterate over ask_posts and append to result_list a list with two elements:
The first element shall be the column created_at.
Because the created_at column is the seventh column in ask_posts, you'll need to get the element at index 6 in each row.
The second element shall be the number of comments of the post.
You'll also need to convert the value to an integer.
Create two empty dictionaries called counts_by_hour and comments_by_hour.
Loop through each row of result_list.
Extract the hour from the date, which is the first element of the row.
Use the datetime.strptime() method to parse the date and create a datetime object.
Use the string we want to parse as the first argument and a string that specifies the format as the second argument.
Use the datetime.strftime() method to select just the hour from the datetime object.
If the hour isn't a key in counts_by_hour:
Create the key in counts_by_hour and set it equal to 1.
Create the key in comments_by_hour and set it equal to the comment number.
If the hour is already a key in counts_by_hour:
Increment the value in counts_by_hour by 1.
Increment the value in comments_by_hour by the comment number.

In [6]:
import datetime as dt

result_list = []

for post in ask_posts:
    result_list.append([post[6], int(post[4])]) #[created at, number of comments]
    
print(result_list[:5])  


[['8/16/2016 9:55', 6], ['11/22/2015 13:43', 29], ['5/2/2016 10:14', 1], ['8/2/2016 14:20', 3], ['10/15/2015 16:38', 17]]


In [7]:
counts_by_hour = {}
comments_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for each_row in result_list:
    comment = each_row[1]
    date = each_row[0]
    time = dt.datetime.strptime(date, date_format).strftime("%H") #1. convert/parse from string to datetime class, 2. format datetime class, extract only the hour %H
    if time in counts_by_hour:
        counts_by_hour[time] += 1
        comments_by_hour[time] += comment
    else:
        counts_by_hour[time] = 1
        comments_by_hour[time] = comment
        
comments_by_hour


{'00': 447,
 '01': 683,
 '02': 1381,
 '03': 421,
 '04': 337,
 '05': 464,
 '06': 397,
 '07': 267,
 '08': 492,
 '09': 251,
 '10': 793,
 '11': 641,
 '12': 687,
 '13': 1253,
 '14': 1416,
 '15': 4477,
 '16': 1814,
 '17': 1146,
 '18': 1439,
 '19': 1188,
 '20': 1722,
 '21': 1745,
 '22': 479,
 '23': 543}

Use the example above to calculate the average number of comments per post for posts created during each hour of the day.
The result should be a list of lists in which the first element is the hour and the second element is the average number of comments per post. Assign the result to a variable named avg_by_hour. Display the results.

In [9]:
avg_by_hour = []

for hr in comments_by_hour:
    avg_by_hour.append([hr, comments_by_hour[hr] / counts_by_hour[hr]])

avg_by_hour

[['03', 7.796296296296297],
 ['17', 11.46],
 ['04', 7.170212765957447],
 ['16', 16.796296296296298],
 ['06', 9.022727272727273],
 ['08', 10.25],
 ['19', 10.8],
 ['12', 9.41095890410959],
 ['22', 6.746478873239437],
 ['00', 8.127272727272727],
 ['11', 11.051724137931034],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['07', 7.852941176470588],
 ['18', 13.20183486238532],
 ['10', 13.440677966101696],
 ['09', 5.5777777777777775],
 ['14', 13.233644859813085],
 ['01', 11.383333333333333],
 ['13', 14.741176470588234],
 ['05', 10.08695652173913],
 ['23', 7.985294117647059],
 ['02', 23.810344827586206],
 ['20', 21.525]]


Use the sorted() function to sort swap_avg_by_hour in descending order. Since the first column of this list is the average number of comments, sorting the list will sort by the average number of comments.
Set the reverse argument to True, so that the highest value in the first column appears first in the list.
Assign the result to sorted_swap.
Print the string "Top 5 Hours for Ask Posts Comments".
Loop through each average and each hour (in this order) in the first five lists of sorted_swap.
Use the str.format() method to print the hour and average in the following format: 15:00: 38.59 average comments per post.
To format the hours, use the datetime.strptime() constructor to return a datetime object and then use the strftime() method to specify the format of the time.
To format the average, you can use {:.2f} to indicate that just two decimal places should be used.
Which hours should you create a post during to have a higher chance of receiving comments? Refer back to the documentation for the data set to convert the times to the time zone you live in. Write a markdown cell explaining your findings.

In [13]:
swap_avg_by_hour = []

for row in avg_by_hour:
    x = row[0]
    y = row[1]
    swap_avg_by_hour.append([y,x])
    
swap_avg_by_hour

sorted_swap = sorted(swap_avg_by_hour, reverse=True)

sorted_swap

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [14]:
# Sort the values and print the the 5 hours with the highest average comments.

print("Top 5 Hours for 'Ask HN' Comments")
for avg, hr in sorted_swap[:5]:
    print(
        "{}: {:.2f} average comments per post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg
        )
    ) # parse -> format
    
    

Top 5 Hours for 'Ask HN' Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post
